File: 백준/Silver/17204. 죽음의 게임/죽음의 게임.py
def dfs(num, cnt):
    global ans
    if num == k:
        ans = min(cnt, ans)
        return
    if not visited[graph[num]]:
        visited[graph[num]] = 1
        dfs(graph[num], cnt + 1)

n, k = map(int, input().split())
ans = 1e9
graph = [0] * (n+1)
visited = [0] * (n+1)
for i in range(n):
    x = int(input())
    graph[i] = x
visited[0] = 1
dfs(graph[0], 1)
if ans == 1e9:
    print(-1)
else:
    print(ans)

File: 백준/Silver/17204. 죽음의 게임/죽음의 게임.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    static int ans;
    static int[] graph;
    static boolean[] visited;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int k = Integer.parseInt(st.nextToken());


        graph = new int[n+1];
        for (int i = 0; i < n; i ++){
            st = new StringTokenizer(br.readLine());
            graph[i] = Integer.parseInt(st.nextToken());
        }

        visited = new boolean[n+1];
        int cnt = 0;
        int num = 0;
        boolean flag = false;
        while(true) {
            cnt ++;
            if (graph[num] == k) {
                break;
            }
            else if (visited[graph[num]]){
                flag = true;
                break;
            }
            else {
                visited[graph[num]] = true;
                num = graph[num];
            }
        }
        if (flag){
            System.out.println(-1);
        }
        else {
            System.out.println(cnt);
        }
    }

}